mxfusion/inference/batch_loop.py

```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
 
         if logger:
             logger.close()
+
+
+class BatchInferenceLoopScipy(GradLoop):
+    """
+    The class for the main loop for batch gradient-based optimization.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._initialized = None
+
+    def run(self, infr_executor, data, param_dict, ctx, optimizer='adam', learning_rate=None, max_iter=1000,
+            verbose=False, logger=None):
+        """
+        :param learning_rate: Not used
+        :param infr_executor: The MXNet function that computes the training objective.
+        :type infr_executor: MXNet Gluon Block
+        :param data: a list of observed variables
+        :type data: [mxnet.ndarray]
+        :param param_dict: The MXNet ParameterDict for Gradient-based optimization
+        :type param_dict: mxnet.gluon.ParameterDict
+        :param ctx: MXNet context
+        :type ctx: mxnet.cpu or mxnet.gpu
+        :param optimizer: the choice of optimizer (default: 'adam')
+        :type optimizer: str
+        :param max_iter: the maximum number of iterations of gradient optimization
+        :type max_iter: int
+        :param verbose: whether to print per-iteration messages.
+        :type verbose: boolean
+        :param logger: The logger to send logs to
+        :type logger: :class:`inference.Logger`
+        """
+
+        if self._initialized is None:
+            self._initialize(param_dict)
+
+        x0 = self._extract_array_from_param_dict(param_dict)
+        func = lambda x: self._f_df(x, infr_executor, data, ctx, param_dict)
```

Comment:
You can directly create an inner function.

Reply:
Do you mean use `def` instead of a lambda?

Reply:
Yes. Then, it is easy to put all the content in ```self._f_df``` into the inner function.